tools/cn-tree-simulator/cn/tree.go

```diff
@@ -0,0 +1,214 @@
+/* SPDX-License-Identifier: Apache-2.0
+ * Copyright (C) 2022 Micron Technology, Inc. All rights reserved.
+ */
+
+package cn
+
+import (
+	"container/list"
+	"encoding/json"
+	"fmt"
+)
+
+type Tree struct {
+	params *TreeParams
+	stats  treeStatistics
+
+	root rootNode
+}
+
+type node struct {
+	children *list.List
+	kvsets   *list.List
+}
+
+type rootNode struct {
+	node
+
+	params *RootNodeParams
+	stats  rootNodeStatistics
+}
+
+type leafNode struct {
+	node
+
+	params *LeafNodeParams
+	stats  leafNodeStatistics
+}
+
+func NewTree(leaves uint, initialSize uint64, params *Params) *Tree {
+	tree := Tree{
+		params: &params.Tree,
+		root: rootNode{
+			node: node{
+				children: list.New(),
+				kvsets:   list.New(),
+			},
+			params: &params.RootNode,
+		},
+	}
+
+	if initialSize != 0 {
+		for i := uint(0); i < leaves; i++ {
+			size := initialSize / uint64(leaves)
+			node := &leafNode{
+				node:   node{kvsets: list.New()},
+				params: &params.LeafNode,
+			}
+
+			tree.root.children.PushBack(node)
+
+			node.kvsets.PushFront(newKvSet(size, false))
+		}
+	}
+
+	return &tree
+}
+
+func (t *Tree) Size() uint64 {
+	total := t.root.size()
+
+	for leaf := t.root.children.Front(); leaf != nil; leaf = leaf.Next() {
+		for kvset := leaf.Value.(*leafNode).kvsets.Front(); kvset != nil; kvset = kvset.Next() {
+			total += kvset.Value.(*kvSet).size
+		}
+	}
+
+	return total
+}
+
+func (t *Tree) Ingest(data uint64) {
+	t.root.kvsets.PushFront(newKvSet(data, false))
+
+	t.stats.ingests++
+	t.stats.ingested += data
+	t.stats.written += data
+
+	if t.root.stats.spills == uint(0) && t.root.size() < t.root.params.InitialSpillSize {
+		return
+	}
+
+	if t.root.size() < t.root.params.SpillSize {
+		return
+	}
```

Comment:
This rule shouldn't apply if `t.root.stats.spills == 0`.   Also, can you check number of children instead of number of spills?  It works out to be the same, but the intent is to base the behavior on number children.  We just say "first spill" in discussion because it easier to say :)

Reply:
A little confused by this comment. Root node shouldn't spill unless it is greater than or equal to spill size correct?

Reply:
Currently spill is triggered by reaching a minimum number of 4 kvsets, and will spill no more than a maximum of 8 kvsets, but i have long believed that we should try to spill based on size.  Say, limit root spills to 8G or 8 kvsets, whichever you find first...  Something like that...

Reply:
Clarifying my comment:  the check at line 91 should only happen the root already has children.  One spill size applies when root has no children,  other spill size applies when it has children.  OR... maybe I'm misreading the code.